hpccm/primitives/arg.py

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# pylint: disable=invalid-name, too-few-public-methods
+
+"""Environment primitive"""
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from __future__ import print_function
+
+import logging # pylint: disable=unused-import
+
+import hpccm.config
+
+from hpccm.common import container_type
+
+class arg(object):
+    """The `arg` primitive sets the corresponding environment
+    variables during the build time of a docker container. 
+    This primitive is docker specific and is ignored for singularity
+
+    # Parameters
+
+    variables: A dictionary of key / value pairs.  The default is an
+    empty dictionary.
+
+    # Examples
+
+    ```python
+    arg(variables={'HTTP_PROXY': 'proxy.example.com', 'NO_PROXY':'example.com'})
+    ```
+    """
+
+    def __init__(self, **kwargs):
+        """Initialize primitive"""
+        self.__variables = kwargs.get('variables', {})
+
+    def __str__(self):
+        """String representation of the primitive"""
+        if self.__variables:
+
+            if hpccm.config.g_ctype == container_type.SINGULARITY or \
```

Comment:
FYI: https://github.com/hpcng/singularity/issues/1676

It is possible to get something like the Dockerfile `ARG` functionality in Singularity with this approach:
`SINGULARITYENV_FOO=bar singularity build foo.sif Singularity.def`

where Singularity.def contains
```
%post
    echo $FOO >> /bar
```

So I wonder if the "nodefault" case is applicable to Singularity?

Reply:
It is possible to simulate that use case for Singularity. But since for singularity is not as clear the difference of ARG and ENV I do not know if it is better to treat the build time variables using the environment primitive